File: c++/leetcode/41. First Missing Positive.cpp
class Solution
{
public:
  int firstMissingPositive(vector<int> &nums)
  {
    if (!nums.size())
      return 1;
    unordered_set<int> seen;
    int min = 1;

    for (auto i : nums)
    {
      if (i > 0)
        seen.insert(i);
    }

    for (auto i : seen)
    {
      if (seen.find(min) != seen.end())
        min++;
      else
        break;
    }

    return min;
  }
};


// correct and O(n)
class Solution
{
public:
  int firstMissingPositive(vector<int> &nums)
  {
    if (nums.empty())
      return 1;
    const int size = nums.size();
    bool seen_one = false;

    // remove invalid : all < 1 and size
    for (int i = 0; i < size; ++i)
    {
      if (nums[i] == 1 && !seen_one)
        seen_one = true;
      else if (nums[i] < 1 || nums[i] > size)
        nums[i] = 1;
    }

    if (!seen_one)
      return 1;

    // mark visited indices
    for (int i = 0; i < size; ++i)
    {
      const int ind = abs(nums[i]) - 1;
      if (nums[ind] > 0)
        nums[ind] *= -1;
    }

    // find missing
    for (int i = 0; i < size; ++i)
    {
      if (nums[i] > 0)
        return i + 1;
    }

    return size + 1;
  }
};

File: javascript/leetcode/41. First Missing Positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function (nums) {
  if (!nums.length) return 1;

  const seen = new Set();
  let min = 1;

  nums.forEach((el) => (el > 0 ? seen.add(el) : null));
  seen.forEach((el) => (seen.has(min) ? min++ : null));

  return min;
};

// correct and O(n)

/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
    if(nums.length === 0) return 1
    
    const size = nums.length
    let seen_one = false
    
    // change invalid to 1 
    nums.forEach((el, ind) => {
        if(el === 1 && !seen_one) seen_one = true
        else if(el < 1 || el > size) nums[ind] = 1
    })  
    
    if(!seen_one) return 1
    
    // mark visited
    nums.forEach(el => {
        const ind = Math.abs(el) - 1
        if(nums[ind] > 0) nums[ind] *= -1
    })  
    
    // find missing
    for(let i = 0; i < size; ++i){
        if(nums[i] > 0) return ++i;
    } 
    
    return size + 1
};  

 
